js/runner.go

```diff
@@ -743,6 +771,31 @@ func (u *VU) runFn(
 	return v, isFullIteration, endTime.Sub(startTime), err
 }
 
+func (u *ActiveVU) incrIteration() {
+	u.iteration++
+	u.state.Iteration = u.iteration
+
+	if u.iterSync != nil {
+		// block other VUs from incrementing scenario iterations
+		u.iterSync <- struct{}{}
+		defer func() {
+			<-u.iterSync // unlock
+		}()
+	}
+
+	if _, ok := u.scenarioIter[u.scenarioName]; ok {
+		u.scenarioIter[u.scenarioName]++
+	} else {
+		u.scenarioIter[u.scenarioName] = 0
+	}
+	if u.getNextScLocalIter != nil {
+		u.scIterLocal = u.getNextScLocalIter()
+	}
+	if u.getNextScGlobalIter != nil {
+		u.scIterGlobal = u.getNextScGlobalIter()
+	}
+}
+
```

Comment:
Why do we not have GlobalIter for some executors?

I do think whenever we have local one we can (and should) have a global one and all executors do have 
1. way to make SegmetnedIndex
2. make iterations somehow 
which just means it needs to call `next` and get the scaled and unscaled values of it ... preferably atomically as I di in [xk6-segment](https://github.com/MStoykov/xk6-segment/blob/8df18d74cb838a7c16167afec6feabb65ae93af0/pkg/segment/segment.go#L112-L122). Even in golang you don't need the additional result type ... it is only useful for javascript.

Reply:
> Why do we not have GlobalIter for some executors?

I think I asked the same question to @na-- since [he mentioned these 3 explicitly in the specs](https://github.com/k6io/k6/pull/1863#issuecomment-782069940), and IIRC he doesn't think it would be useful for the other executors. I do agree that it would make things neater if we could just define this on `BaseExecutor`, but honestly I don't mind the current state much.

I don't quite understand what would be the benefit of changing `next`. The only difference I can see in the `xk6-segment` implementation is that we would avoid the additional read lock by not having to call `GetUnscaled()`. But semantically it makes more sense to me for `Next()` to only increment and to have a separate getter. It's also more flexible to use and I would have to change how it's used in `ramping-vus`.

Reply:
I would like to get @na-- to review this PR before I continue commenting after this PR. I think I am pulling this PR in a different direction than what you think it should be and would like to get a third opinion if I or you are right on the different issues or if ... we are both wrong :rofl: 

Regarding this ... I think you keep slapping synchronization on top of things trying to fix them not synchronizing , while I am trying to see if we can do with the amount of synchronization we have currently. Some thoughts:

- SegmentedIndex doesn't need to be thread-safe for `ramping-vus`, so it means that its performance will suffer (somewhat) for something it doesn't need
- The places where SegementedIndex needs to be thread-safe are the same places where it needs to be threadsafe from `next`/`prev`/`GoTo` to `GetScaled` and `GetUnscaled` (might've gotten a name wrong, I ain't looking at the code at the moment).  For me, this means that there should be `SyncSegmentedIndex`(better name exist) that returns the `scaled`, `unscaled` in the cases where it needs to. Even if we only care about `next` we can probably just have an even smaller wrapper for now.
- Again you already lock for `next` and then ... you lock once more around it (through additional means) to get the 2 counters you care about, so combined with the below I think you can have it all done at once with *1* synchronization mechanism wrapping a nonsynchronized SegmentedIndex which can still be used in the cases where synchronization ... isn't required.
- I see no reason why we shouldn't have iteration counters for all executors ... we might want them only for specific ones, but there is nothing wrong (and it definitely will be useful) for even the externally-controlled one to have them so they can do the same tricks the `shared-iterations` can do. And given that the current implementation clearly will be simpler if all of them have it instead of only some - I am for all of them having it.

Reply:
> > Why do we not have GlobalIter for some executors?
> 
> I think I asked the same question to @na-- since [he mentioned these 3 explicitly in the specs](https://github.com/k6io/k6/pull/1863#issuecomment-782069940), and IIRC he doesn't think it would be useful for the other executors.

I am currently reviewing this PR and still have a lot of things I need to understand in it, but if you're referencing this sentence of mine:
> for `shared-iterations`, `constant-arrival-rate` and `variable-arrival-rate` scenarios, a way to get the sequential number of the current iteration among _all_ of the iterations for the current scenario, across all VUs and even instances.

Then the reason for listing only these 3 executors is simple - they are the only ones where we can actually _know_ the sequential number of an iteration globally, among all VUs and instances, without any synchronization. This is because the iteration number is the thing we segment on, not VUs, as is the case with the rest of the executors.

Reply:
I am still reviewing this (getting easily distracted and finding it a bit difficult to make headway :disappointed: ), but yeah, as @MStoykov mentioned above, something in the synchronization looks a bit wrong or unnecessary.

I think his first point is about the fact that `SegmentedIndex.Next()` has locking: https://github.com/k6io/k6/blob/25ee787380f7917ca33115711612a6a6d724f29f/lib/execution_segment.go#L750-L758
But we also lock where it's used, for example: https://github.com/k6io/k6/blob/25ee787380f7917ca33115711612a6a6d724f29f/lib/executor/shared_iterations.go#L193-L197

If `Next()` (and the other traversal methods) returned the new scaled and unscaled values, you wouldn't need to lock twice and ` GetScaled()` and `GetUnscaled()` will also be obsolete, I think?

Reply:
yes, that is what I meant.

Reply:
OK, I'll try refactoring this then.

Reply:
Take a look at e9feff31. I decided to keep returning a struct instead of multiple values, to simplify future expansion.

Reply:
The other comment here is about adding global iterations to all executors. @na-- WDYT?

Reply:
https://github.com/k6io/k6/pull/2057 what I actually meant ... I think've missed initially that you have the local iteration as a counter that you atomically increment. 

The commit doesn't fix any comments and the names are not great, but I think that + the commit message should be good enough explanation